File: dotnet-resend-examples/Examples/Segments.cs
namespace ResendExamples;

using Resend;

public static class Segments
{
    public static async Task RunAsync()
    {
        var apiKey = Environment.GetEnvironmentVariable("RESEND_API_KEY")
            ?? throw new Exception("RESEND_API_KEY environment variable is required");

        var client = ResendClient.Create(apiKey);

        // 1. List segments
        Console.WriteLine("=== Listing Segments ===");
        var segments = await client.SegmentListAsync();
        foreach (var segment in segments.Content.Data)
        {
            Console.WriteLine($"  - {segment.Name} ({segment.Id})");
        }

        // 2. Create a contact
        // Contacts are account-level in the .NET SDK: they are not scoped to a
        // segment, so assigning a contact to a segment is a separate call (step 3).
        Console.WriteLine("\n=== Creating Contact ===");
        var contact = await client.ContactAddAsync(new ContactData
        {
            Email = "clicked@resend.dev",
            FirstName = "Jane",
            LastName = "Doe",
            IsUnsubscribed = false
        });
        var contactId = contact.Content;
        Console.WriteLine($"Contact created: {contactId}");

        // 3. Add the contact to a segment
        var segmentIdValue = Environment.GetEnvironmentVariable("RESEND_SEGMENT_ID");
        if (!string.IsNullOrEmpty(segmentIdValue) && Guid.TryParse(segmentIdValue, out var segmentId))
        {
            Console.WriteLine("\n=== Adding Contact to Segment ===");
            await client.ContactAddToSegmentAsync(contactId, segmentId);
            Console.WriteLine($"Contact {contactId} added to segment {segmentId}");
        }
        else
        {
            Console.WriteLine("\nSkipping segment assignment: set RESEND_SEGMENT_ID to a valid segment id to try this step.");
        }

        // 4. List contacts
        Console.WriteLine("\n=== Listing Contacts ===");
        var contacts = await client.ContactListAsync();
        foreach (var c in contacts.Content.Data)
        {
            Console.WriteLine($"  - {c.FirstName} {c.LastName} <{c.Email}> (unsubscribed: {c.IsUnsubscribed})");
        }

        // 5. Update the contact
        Console.WriteLine("\n=== Updating Contact ===");
        await client.ContactUpdateAsync(contactId, new ContactData
        {
            FirstName = "Janet",
            IsUnsubscribed = false
        });
        Console.WriteLine("Contact updated: Jane -> Janet");

        // 6. Remove the contact
        Console.WriteLine("\n=== Removing Contact ===");
        await client.ContactDeleteAsync(contactId);
        Console.WriteLine($"Contact removed: {contactId}");

        Console.WriteLine("\nDone! Full segment/contact lifecycle complete.");
    }
}

File: python-resend-examples/examples/segments.py
#!/usr/bin/env python3
"""
Segments & Contacts Management

Demonstrates managing segments (contact lists) and contacts
using the Resend API.

Usage:
    python examples/segments.py

See: https://resend.com/docs/api-reference/segments
"""

import os
import resend
from dotenv import load_dotenv

load_dotenv()

resend.api_key = os.environ["RESEND_API_KEY"]

# Use existing segment or create one in the dashboard
segment_id = os.environ.get("RESEND_SEGMENT_ID", "your-segment-id")

print("=== Segments & Contacts Management ===\n")

# List all segments
print("Listing segments...")
segments = resend.Segments.list()
print(f"Found {len(segments.get('data', []))} segment(s)")
for segment in segments.get("data", []):
    print(f"  - {segment['name']} ({segment['id']})")
print()

# Add a contact to a segment
print("Adding contact to segment...")
contact = resend.Contacts.create({
    "email": "clicked@resend.dev",
    "first_name": "Jane",
    "last_name": "Doe",
    "unsubscribed": False,
    "segments": [{"id": segment_id}],
})
print(f"Contact created: {contact['id']}")
print()

# List contacts in the segment
print("Listing contacts in segment...")
contacts = resend.Contacts.list(segment_id=segment_id)
print(f"Found {len(contacts.get('data', []))} contact(s)")
for c in contacts.get("data", [])[:5]:
    print(f"  - {c['email']} ({c.get('first_name', '')} {c.get('last_name', '')})")
print()

# Update a contact
print("Updating contact...")
updated = resend.Contacts.update({
    "id": contact["id"],
    "first_name": "Janet",
    "unsubscribed": False,
})
print(f"Contact updated: {updated['id']}")
print()

# Remove a contact
print("Removing contact...")
resend.Contacts.remove(id=contact["id"])
print("Contact removed successfully")
